### ISSUE TITLE
Cannot craft ender eyes

### ISSUE THREAD

**Issue by JamesGreen31** [gaucho-matrero/altoclef#285]
development issue for issue #279 

## Expected Behavior

Bot makes blaze powder
Bot makes eyes of ender

## Actual Behavior (What lead to the bot getting permanently stuck)

Bot swaps between the tasks as it thinks that blaze powder is achieved before it actually is

## Crashlogs and Screenshots (if applicable)
_(you may drag + drop text logs and images here, thanks GitHub)_

` private Task getEyesOfEnderTask(AltoClef mod, int targetEyes) {
        if (mod.getEntityTracker().itemDropped(Items.ENDER_EYE)) {
            setDebugState("Picking up Dropped Eyes");
            return new PickupDroppedItemTask(Items.ENDER_EYE, targetEyes);
        }

        int eyeCount = mod.getItemStorage().getItemCount(Items.ENDER_EYE);

        int blazePowderCount = mod.getItemStorage().getItemCount(Items.BLAZE_POWDER); // THIS RETURNS 14 BEFORE THE CRAFTING IS COMPLETE
        int blazeRodCount = mod.getItemStorage().getItemCount(Items.BLAZE_ROD);
        int blazeRodTarget = (int)Math.ceil(((double)targetEyes - eyeCount - blazePowderCount) / 2.0);
        int enderPearlTarget = targetEyes - eyeCount;
        boolean needsBlazeRods = blazeRodCount < blazeRodTarget;
        boolean needsBlazePowder = eyeCount + blazePowderCount < targetEyes;
        boolean needsEnderPearls = mod.getItemStorage().getItemCount(Items.ENDER_PEARL) < enderPearlTarget;

        if (needsBlazePowder && !needsBlazeRods) {
            // We have enough blaze rods.
            setDebugState("Crafting blaze powder");
            return TaskCatalogue.getItemTask(Items.BLAZE_POWDER, targetEyes - eyeCount);
        }

        if (!needsBlazePowder && !needsEnderPearls) {
            // Craft ender eyes
            setDebugState("Crafting Ender Eyes");
            return TaskCatalogue.getItemTask(Items.ENDER_EYE, targetEyes);
        }//CAUSES CYCLING BETWEEN THESE TWO IF BLOCKS`

**Comment by JamesGreen31** [comment]
Issues are written in commetns in all caps

**Comment by JamesGreen31** [comment]
Attempted fix num 1)

Duct Tape solution:
    Make sure that the make powder task is fully done before moving on. Hopefully, this will get it out of the inventory before 
    something else takes over.

    In the future, we should figure out why the 2x2 grid is counting potential results as inventory.

**Comment by JamesGreen31** [comment]
attempted fix #2:
                check if we are making eyes and then don't stop until we have them.

**Comment by JamesGreen31** [comment]
attempted fix 3)
    Swap crafting ender eyes with blaze powder. i.e, make eyes before powder.

**Comment by JamesGreen31** [comment]
Renamed as the get items count is not the issue

**Comment by JamesGreen31** [comment]
Attempted fix 4)
                    Try crafting eyes first. This will force that task to finish before worrying about rods

**Comment by JamesGreen31** [comment]
Observation:
                    If I provide the bot all the items it needs, such as iron swords, it moves to "collecting 11 beds".
                    There may not actually be an issue here, but the order of the tasks may be broken. Meaning, if the
                    bot does not have 2 iron picks and 1 sword by the time its making eyes, it will break.

**Comment by JamesGreen31** [comment]
I think that the "Ensure 2x2 crafting grid" is being called somewhere errantly
                    as we narrowed it down to simply being stuck clearing the grid

**Comment by JamesGreen31** [comment]
Observation
    we discovered that blaze powder count is wrong

**Comment by JamesGreen31** [comment]
crafting eneder eyes manually (placing ingrediants in a chest, telling alto to make eyes) works. Error is certainly located in BM2

**Comment by JamesGreen31** [comment]
![image](https://user-images.githubusercontent.com/72313560/183322604-e1313a93-7586-4e9f-8809-54fff6af9488.png)
When blaze powder is in the crafting grid, the count is 0

**Comment by JamesGreen31** [comment]
attempted fix 4)
    Check specifically for the crafting slots. Not sure why these were not included before

**Comment by JamesGreen31** [comment]
the crafting fails because "got powder" relies on the current count of blaze powder, which because of the structure of beat minecraft, is obtained each frame during crafting. Because we intentionally do not consider items in the crafting grid as "obtained", the count for blaze powder goes to 0 once the powder is added as part of the recipe. This changes the "got powder" boolean to false when it should be true resulting in the bot clearing its inventory to "get the powder". Once it has the powder, the loop restarts.

This problem is located in the inventory sub tracker

`   private static boolean shouldIgnoreSlotForContainer(Slot slot) {
        // IMPORTANT NOTE:!!!!
        // Ignore crafting table output when calculating container slots.
        //
        // Why?
        // Because we don't want the bot to think we "have" an item if it's in our output slot. Otherwise it will
        // softlock because it will assume we're all good (we got the item!) when in reality we need to grab that item.
        //
        // We also don't want our bot to think we "have" an item if it's in our armor/crafting/shield slots. That's annoying to work with. //TODO re-consider this
        if (slot instanceof CraftingTableSlot && slot.equals(CraftingTableSlot.OUTPUT_SLOT))
            return true;
        if (slot instanceof PlayerSlot) {
            // Ignore non-normal inventory slots
            int window = slot.getWindowSlot();
            return window < 9 || window > 44;
        }
        return false;
    }`

**Comment by JamesGreen31** [comment]
Changing the above may cause issues. In short, by not counting these items, we cannot perform logical steps in between crafting phases. However, it isn't an issue with crafting table slots. Notice that for crafting tables, we care if it is the output slot, but not input slots.

Since the bot **no longer carries items in its 2x2 crafting grid**, we should count items in there.

**Comment by JamesGreen31** [comment]
Partial Progress
    Changing the inventory sub tracker to include the count of items in the player crafting slot fixes the crafting issue, but results in the items being stuck in the slot

**Comment by JamesGreen31** [comment]
**Solution write up**
In order to craft properly in @gamer we have to ensure our count is correct throughout the execution. This means that we had to count items in the crafting grid (2x2). However, this stood in conflict with the estabolished schema of "don't count items in the crafting grid". As a result, when this was changed, the bot made attempts to store items in the crafting grid as it normally would. A few updates ago, it was decided that we should not use the 2x2 grid as item storage and wrote an interface ITaskRequiresCrafting which would clear the crafting grid anytime the grid was not specifically being used for crafting. This caused a loop as Alto would store an item in the crafting grid, pick it up to clear the grid, and then store it there again. A simple change was made to inventory tracker that prohibits alto from considering the 2x2 grid as a storage solution. This resolved the issue and improved run time across the board as altoclef spends considerably less time clearing the grid now since items are never placed there unless specifically for crafting.
